akka-remote/src/main/scala/akka/remote/artery/tcp/ssl/ConfigSSLEngineProvider.scala

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright (C) 2018-2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.remote.artery.tcp.ssl
+
+import java.security.SecureRandom
+
+import akka.actor.ActorSystem
+import akka.event.Logging
+import akka.event.MarkerLoggingAdapter
+import akka.remote.artery.tcp.SSLEngineProvider
+import akka.remote.artery.tcp.SecureRandomFactory
+import com.typesafe.config.Config
+import javax.net.ssl.SSLContext
+import javax.net.ssl.SSLEngine
+import javax.net.ssl.SSLSession
+
+/**
+ * Config in akka.remote.artery.ssl.config-ssl-engine
+ */
+final class ConfigSSLEngineProvider(protected val config: Config, protected val log: MarkerLoggingAdapter)
+    extends SSLEngineProvider {
+
+  def this(system: ActorSystem) =
+    this(
+      system.settings.config.getConfig("akka.remote.artery.ssl.config-ssl-engine"),
+      Logging.withMarker(system, classOf[ConfigSSLEngineProvider].getName))
+
+  private val keyStoreProviders = new JksManagersProvider(config)
+  private val rng: SecureRandom = SecureRandomFactory.createSecureRandom(config, log)
+  private lazy val sslFactory: SslFactory = new SslFactory(config, keyStoreProviders, rng)(log)
+  private lazy val sslContext: SSLContext = sslFactory.sslContext
+
+  override def createServerSSLEngine(hostname: String, port: Int): SSLEngine =
+    sslFactory.createServerSSLEngine(hostname, port)
+
+  override def createClientSSLEngine(hostname: String, port: Int): SSLEngine =
+    sslFactory.createClientSSLEngine(hostname, port)
+
+  override def verifyClientSession(hostname: String, session: SSLSession): Option[Throwable] =
+    NoopSessionVerifier.verifyClientSession(hostname, session)
+
+  override def verifyServerSession(hostname: String, session: SSLSession): Option[Throwable] =
+    NoopSessionVerifier.verifyServerSession(hostname, session)
+
+}
```

Comment:
@patriknw as you pointed out, there were multiple places where I was introducing issues if users have opted into overriding methods or accessing fields I have removed from the old `ConfigSSLEngineProvider`. Instead, I've rolled back the original implementation but deprecated it. Then, I've introduced a this new one, also named `ConfigSSLEngineProvider`, but moved to the `akka.remote.artery.tcp.ssl` where the rest of the code of this PR resides. It is also `final`.

The old and new `ConfigSSLEngineProvider` implementations are interchangeable since both read the exact same settings and none adds any extra behavior (e.g. caching or authorization). Users overriding methods on the old `ConfigSSLEngineProvider` were already setting up the FQCN of their class name in settings so I think it's even safe to replace the following in `reference.conf`:

```diff
- ssl-engine-provider = akka.remote.artery.tcp.ConfigSSLEngineProvider
+ ssl-engine-provider = akka.remote.artery.tcp.ssl.ConfigSSLEngineProvider
```

Reply:
That looks like a very good approach. Then we can phase out the old deprecated and have a brighter future.